embassy-nrf/src/qdec.rs

```diff
@@ -0,0 +1,222 @@
+        a: AnyPin,
+        b: AnyPin,
+        led: Option<AnyPin>,
+        config: Config,
+    ) -> Self {
+        unborrow!(irq);
+        let r = Self::regs();
+
+        // Select pins.
+        a.conf().write(|w| w.input().connect().pull().pullup());
+        b.conf().write(|w| w.input().connect().pull().pullup());
+        r.psel.a.write(|w| unsafe { w.bits(a.psel_bits()) });
+        r.psel.b.write(|w| unsafe { w.bits(b.psel_bits()) });
+        if let Some(led_pin) = &led {
+            led_pin.conf().write(|w| w.dir().output());
+            r.psel.led.write(|w| unsafe { w.bits(led_pin.psel_bits()) });
+        }
+
+        // Enables/disable input debounce filters
+        r.dbfen.write(|w| match config.debounce {
+            true => w.dbfen().enabled(),
+            false => w.dbfen().disabled(),
+        });
+
+        // Set LED output pin polarity
+        r.ledpol.write(|w| match config.led_polarity {
+            LedPolarity::ActiveHigh => w.ledpol().active_high(),
+            LedPolarity::ActiveLow => w.ledpol().active_low(),
+        });
+
+        // Set time period the LED is switched ON prior to sampling (0..511 us).
+        r.ledpre
+            .write(|w| unsafe { w.ledpre().bits(config.led_pre_usecs.min(511)) });
+
+        // Set sample period
+        r.sampleper.write(|w| match config.period {
+            SamplePeriod::_128us => w.sampleper()._128us(),
+            SamplePeriod::_256us => w.sampleper()._256us(),
+            SamplePeriod::_512us => w.sampleper()._512us(),
+            SamplePeriod::_1024us => w.sampleper()._1024us(),
+            SamplePeriod::_2048us => w.sampleper()._2048us(),
+            SamplePeriod::_4096us => w.sampleper()._4096us(),
+            SamplePeriod::_8192us => w.sampleper()._8192us(),
+            SamplePeriod::_16384us => w.sampleper()._16384us(),
+            SamplePeriod::_32ms => w.sampleper()._32ms(),
+            SamplePeriod::_65ms => w.sampleper()._65ms(),
+            SamplePeriod::_131ms => w.sampleper()._131ms(),
+        });
+
+        // Enable peripheral
+        r.enable.write(|w| w.enable().set_bit());
+
+        irq.disable();
+        irq.set_handler(|_| {
+            let r = Self::regs();
+            r.intenclr.write(|w| w.reportrdy().clear());
+            WAKER.wake();
+        });
+        irq.enable();
+
+        Self {
+            phantom: PhantomData,
+        }
+    }
+
+    /// Perform an asynchronous read of the decoder.
+    /// The returned future can be awaited to obtain the number of steps.
+    ///
+    /// If the future is dropped, the read is cancelled.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// let irq = interrupt::take!(QDEC);
+    /// let config = qdec::Config::default();
+    /// let mut q = Qdec::new(p.QDEC, p.P0_31, p.P0_30, config);
+    /// let delta = q.read().await;
+    /// ```
+    pub async fn read(&mut self) -> i16 {
+        // In case the future is dropped, stop the task and reset events.
```

Comment:
Is this needed? I think if you stop QDEC then it'll stop counting. If rotation happens while hte user is not reading, it'll be lost.

It might be useful to `select()` between QDEC read and a timeout, in which case the user might want to read later and not lose the rotation count.